keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,707 @@
+
+Every VM comes with a management related overhead which can sporadically be pretty significant (memory streaming, SRIOV attachment, gpu attachment, virtio-fs, …). Swap helps to not request much more memory to deal with short term worst-case scenarios.
+
+With virtualization, clusters are typically provisioned based on the workloads’ memory consumption, and any infrastructure container overhead is overcommitted. This overhead could be safely swapped out.
+
+- Required for live migration of VMs https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-754878431
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+### Notes/Constraints/Caveats (Optional)
+
+In changing the CRI, we must ensure that container runtime downstreams are able to support the new configurations.
+
+We considered adding parameters for both per-workload `memory-swap` and `swappiness`. These are documented as part of the Open Containers [runtime specification] for Linux memory configuration. Since `memory-swap` is a per-workload parameter, and `swappiness` is optional and can be set globally, we are choosing to only expose `memory-swap` which will adjust swap available to workloads.
+
+Since we are not currently setting `memory-swap` in the CRI, the default behaviour is to allocate the same amount of swap for a workload as memory requested. We will update the default to not permit the use of swap by setting `memory-swap` equal to `limit`.
+
+[runtime specification]: https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+
+### Risks and Mitigations
+
+Having swap available on a system reduces predictability. Swap's performance is worse than regular memory, sometimes by many orders of magnitude, which can cause unexpected performance regressions. Furthermore, swap changes a system's behaviour under memory pressure, and applications cannot directly control what portions of their memory usage are swapped out. Since enabling swap permits greater memory usage for workloads in Kubernetes that cannot be predictably accounted for, it also increases the risk of noisy neighbours and unexpected packing configurations, as the scheduler cannot account for swap memory usage.
+
+This risk is mitigated by preventing any workloads from using swap by default, even if swap is enabled and available on a system. This will allow a cluster administrator to test swap utilization just at the system level without introducing unpredictability to workload resource utilization.
+
+Additionally, we will mitigate this risk by determining a set of metrics to quantify system stability and then gathering test and production data to determine if system stability changes when swap is available to the system and/or workloads in a number of different scenarios.
+
+Since swap provisioning is out of scope of this proposal, this enhancement poses low risk to Kubernetes clusters that will not enable swap.
+
```

Comment:
If you set an `emptyDir` volume to be of type `Memory` and also enable a pagefile, can that `emptyDir` data be written to nonvolatile storage? If so, does this introduce any risk?

Reply:
This scenario isn't really covered by our swap scenarios/user stories; can you think of one it might fall under? For alpha we are only supporting swap partitions, not swap files.

Reply:
_As a_ developer _I want to provide some data from a Secret to a Pod
and I want to ensure that those secret data are never written to nonvolatile storage in plaintext
so that I can demonstrate compliance with an externally-defined standard_

_As a_ criminal _I want to steal some hardware
and make a copy of the data previously written to nonvolatile storage
so that I can misuse the secrets I recover_

This implies I must also configure a form of encryption for swap, or not use swap at all. Previously I could rely on using `emptyDir` of type `Memory`; however, that “Memory” designation only holds if the Linux system that runs the node is configured not to use any kind of pagefile. It doesn't matter whether there is a swap device or a swap file, _tmpfs_ can be flushed to swap in either case.

If Kubernetes instead used _ramfs_ for this kind of `emptyDir` then the story would be different. AIUI _ramfs_ is only ever held in actual memory.

Reply:
(this might be an area where SIG Security can provide a recommendation and / or confirm the impact)

Reply:
I would say this scenario is not applicable to this KEP. Given the constraints, if contents of memory should never be swapped out for compliance reasons, then swap should not be enabled on the system.

Reply:
My concern is that we're potentially switching from a case where running with swap enabled is hard to one where running with swap enabled is feasible, and the person turning on swap might be focused on (say) cost or performance rather than information security.

I would definitely want to add a warning everywhere we document `--fail-swap-on`, so that readers are aware of its potential impact on confidentiality.

(yes this is initially alpha, but if we provide the switch then I'm worried folks will press it)